src/UpdateChannel.js

```diff
@@ -49,83 +105,16 @@ export function issueMutation(mutation: Mutation, options: { [key: string]: bool
     new Id(mutation.target, 'local-' + (localCount++));
 
   if (!options.waitForServer) {
-    // Set up the optimistic mutation
-    var subscribers = [];
-    var updates;
-    var latest;
-
-    if (mutation.action === 'CREATE') {
-      executionId = ObjectStore.stackMutation(target, mutation);
-      latest = ObjectStore.getLatest(target);
-      updates = {
-        id: target,
-        latest: latest,
-        fields: Object.keys(latest)
-      };
-    } else {
-      executionId = ObjectStore.stackMutation(target, mutation);
-      subscribers = ObjectStore.fetchSubscribers(target);
-      if (mutation.action === 'DESTROY') {
-        updates = {
-          id: target,
-          latest: null,
-          fields: []
-        };
-      } else {
-        latest = ObjectStore.getLatest(target);
-        updates = {
-          id: target,
-          latest: latest,
-          fields: Object.keys(latest)
-        };
-      }
-    }
-
-    // Push the latest object to matching queries
-    pushUpdates(subscribers, updates);
+    return performOptimisticMutation(target, mutation, options.batch);
   }
 
   var p = new Parse.Promise();
-  MutationExecutor.execute(
-    mutation.action,
-    mutation.target,
-    mutation.data,
-    options.batch
-  ).then(function(result) {
-    var changes;
+  MutationExecutor.execute(mutation, options.batch).then((result) => {
     var subscribers = ObjectStore.fetchSubscribers(target);
     var delta = mutation.generateDelta(result);
-    if (!options.waitForServer) {
-      // Replace the current entry with a Delta
-      changes = ObjectStore.resolveMutation(target, executionId, delta);
-      p.resolve(pushUpdates(subscribers, changes));
-    } else {
-      // Apply it to the data store
-      changes = ObjectStore.commitDelta(delta);
-      p.resolve(pushUpdates(subscribers, changes));
-    }
-  }, function(err) {
-    if (!options.waitForServer) {
-      // Roll back optimistic changes by deleting the entry from the queue
-      var subscribers = ObjectStore.fetchSubscribers(target);
-      if (mutation.action === 'CREATE') {
-        // Make sure the local object is removed from any result sets
-        for (var i = 0; i < subscribers.length; i++) {
-          var subscriber = SubscriptionManager.getSubscription(subscribers[i]);
-          subscriber.removeResult(target);
-        }
-        ObjectStore.destroyMutationStack(target);
-      } else {
-        var noop = new Delta(target, {});
-        var changes = ObjectStore.resolveMutation(
-          target,
-          executionId,
-          noop
-        );
-        pushUpdates(subscribers, changes);
-      }
-    }
-    p.reject(err);
+    // Apply it to the data store
+    var changes = ObjectStore.commitDelta(delta);
+    p.resolve(pushUpdates(subscribers, changes));
   });
 
   return p;
```

Comment:
This promise will never be rejected if we choose to wait for the server and an error occurs.

It's probably cleanest to
```js
return MutationExecutor.execute(mutation, options.batch).then((result) => {
  var subscribers = ObjectStore.fetchSubscribers(target);
  var delta = mutation.generateDelta(result);
  var changes = ObjectStore.commitDelta(delta);
  return pushUpdates(subscribers, changes);
});
```

That way we can return the original promise for chaining, and pass through the update data on success.